src/shim/shim.c

```diff
@@ -139,6 +140,34 @@ static void _shim_load() {
         logger_setDefault(shimlogger_new(log_file));
     }
 
+    // Ensure that the child process exits when Shadow does.  Shadow
+    // ought to have already tried to terminate the child via SIGTERM
+    // before shutting down (though see
+    // https://github.com/shadow/shadow/issues/903), so now we jump all
+    // the way to SIGKILL.
+    if (prctl(PR_SET_PDEATHSIG, SIGKILL) < 0) {
+        warning("prctl: %s", strerror(errno));
+    }
+
+    const char* shadow_pid_str = getenv("SHADOW_PID");
+    if (shadow_pid_str) {
+        unsigned long long tmp_pid = 0;
+        pid_t shadow_pid = 0;
+        int rc = sscanf(shadow_pid_str, "%llu", &tmp_pid);
+        if (rc != 1) {
+            error("SHADOW_PID does not contain an unsigned: %s", shadow_pid_str);
+        } else {
+
+            assert((pid_t)tmp_pid == tmp_pid);
```

Comment:
Is this vacuously true?

Reply:
I think it's to validate it wasn't truncated in the conversion to pid_t? +1 that it could use a comment

It'd be nice if we could parse into a pid_t directly, but I suppose that's tricky since we can't know the format specifier to provide to sscanf (without assuming something about the definition of pid_t).

Reply:
OTOH we could just not do this downcast, and compare directly. C will widen the pid if it's narrower.
```
unsigned long long shadow_pid = 0;
if (sscanf(shadow_pid_str, "%llu", &shadow_pid) != 1) {
...
// Implicitly widen return value of getppid
if (getppid() != shadow_pid) {
```

Btw this control flow would probably be a bit easier to follow by using early exits to avoid excessive nesting.
```
if (/* something failed */) {
  /* log and exit */
}
if (/* something else failed */) {
  /* log and exit */
}
```

Reply:
Fixed